ckanext/harvester_dashboard/blueprints/dashboard.py

```diff
@@ -0,0 +1,160 @@
+    c.job_result = request.params.get("job_result", RESULT_ALL)
+    c.job_run = request.params.get("job_run", RESULT_ALL)
+    context = {"user": c.user, "auth_user_obj": c.userobj}
+    harvest_source_list = tk.get_action("get_harvest_source_infos_for_user")(
+        context, {}
+    )  # noqa
+    c.source_type_options = _get_source_type_options(harvest_source_list)
+    c.job_result_options = RESULT_OPTIONS
+    c.job_run_options = RUN_OPTIONS
+    harvest_source_list = list(
+        filter(
+            lambda harvest_source_info: _source_type_test(
+                harvest_source_info, c.source_type
+            ),
+            harvest_source_list,
+        )
+    )
+    harvest_source_list = list(
+        filter(
+            lambda harvest_source_info: _job_result_test(
+                harvest_source_info, c.job_result
+            ),
+            harvest_source_list,
+        )
+    )
+    harvest_source_list = list(
+        filter(
+            lambda harvest_source_info: _job_run_test(harvest_source_info, c.job_run),
+            harvest_source_list,
+        )
+    )
+    if c.q:
+        harvest_source_list = list(
+            filter(
+                lambda harvest_source_info: _source_name_test(harvest_source_info, c.q),
+                harvest_source_list,
+            )
+        )
+    c.harvest_source_infos = harvest_source_list
+    return render("harvester_dashboard/list.html")
+
+
+def _source_type_test(harvest_source_info, source_type):
+    if source_type == RESULT_ALL:
+        return True
+    source_type_match = (
+        harvest_source_info.get("source")
+        and harvest_source_info.get("source").type == source_type
+    )
+    if source_type_match:
+        return True
+    return False
+
+
+def _source_name_test(harvest_source_info, q):
+    organization_title = harvester_dashboard_organization_title(
+        harvest_source_info.get("organization").title
+    )
+    organization_match = q in organization_title
+    harvest_source_match = q in harvest_source_info.get("source").title
+    if organization_match or harvest_source_match:
+        return True
+    return False
+
+
+def _job_result_test(harvest_source_info, job_result):
+    if job_result == RESULT_ALL:
+        return True
+    job = harvest_source_info.get("job")
+    if job:
+        stats = job.last_job.get("stats")
+        if job_result == RESULT_GATHER_ERRORS:
+            if job.last_job.get("gather_error_summary"):
+                return True
+            else:
+                return False
+        if job_result == RESULT_NO_UPDATES:
+            if stats and not stats.get("updated") and not stats.get("added"):
+                return True
+            else:
+                return False
+        if job_result == RESULT_IMPORT_ERRORS:
+            if job.last_job.get("object_error_summary"):
+                return True
+            else:
+                return False
+        if job_result == RESULT_OK:
```

Comment:
can we rewrite this part, it is hard to read, at leas for me
is that what you mean? 
`if not stats:
    return False
if stats.get("updated") or stats.get("added"):
    return True
if stats.get("errored"):
    return False
return True`

Reply:
I agree that this is hard to read. I only copied the old code and tried to migrate it to Python 3. I will check if I can refactor it with little effort without breaking anything, otherwise I would rather look at this after the general update 😅